### PR TITLE
Complex-valued variance scaling initializers

### PR DESCRIPTION
**Pull request by wdphy16**
Implements complex dtype in `jax.nn.initializers.variance_scaling`, as proposed in #5312 (new discussions after August 2021). It's a necessary step to make Flax layers work with complex numbers in an intuitive way (google/flax#805), with the smallest changes possible.

The initialization of complex variables is discussed in [C. Trabelsi, et al., Deep Complex Networks, 2017](https://arxiv.org/abs/1705.09792) and widely cited after that. They proposed an axisymmetric distribution implemented by first sampling the modulus from the radial CDF, then uniformly sampling the phase. We followed this idea to implement truncated-normal and uniform distributions.

The docstring is adapted from [TensorFlow](https://github.com/tensorflow/tensorflow/blob/9820787fa2db8efbd7f1bfcba72b9af07e5c858b/tensorflow/python/ops/init_ops_v2.py#L499).

cc @PhilipVinc

### CODE REVIEW

**Comment by wdphy16** [comment]
Sorry that I didn't test with `JAX_ENABLE_X64=1`. Let me fix that...

File: jax/_src/nn/initializers.py [hunk 1]
```diff
@@ -48,7 +48,67 @@ def _compute_fans(shape: core.NamedShape, in_axis=-2, out_axis=-1):
   fan_out = shape[out_axis] * receptive_field_size
   return fan_in, fan_out
 
+def _get_real_dtype(dtype):
```
Comment (wdphy16) [inline jax/_src/nn/initializers.py:51]: Is there already an util function in the jax codebase to do this?

Comment (jakevdp) [reply jax/_src/nn/initializers.py:51]: It's definitely done elsewhere in the package, but perhaps a one-liner is better than a utility routine? e.g. https://github.com/google/jax/blob/b6e25fa00c31c12cc4d302e80a44baf6fee3327d/jax/_src/random.py#L677

Comment (jakevdp) [reply jax/_src/nn/initializers.py:51]: Here's another version: https://github.com/google/jax/blob/97a5719fcb40af7231b5f803f965063538282f8e/jax/_src/lax/fft.py#L63

File: jax/_src/nn/initializers.py [hunk 2]
```diff
@@ -48,7 +48,67 @@ def _compute_fans(shape: core.NamedShape, in_axis=-2, out_axis=-1):
...
+  if dtype == jnp.complex64:
+    return jnp.float32
+  elif dtype == jnp.complex128:
+    return jnp.float64
+  else:
+    return dtype
+
+def _complex_uniform(key, shape, dtype):
```
Comment (jekbradbury) [inline jax/_src/nn/initializers.py:59]: It's possible that one or both of these should eventually live in jax.random, but here is a reasonable place for them for now.

File: jax/_src/nn/initializers.py [hunk 3]
```diff
@@ -59,16 +119,25 @@ def init(key, shape, dtype=dtype):
...
     if distribution == "truncated_normal":
-      # constant is stddev of standard normal truncated to (-2, 2)
-      stddev = jnp.sqrt(variance) / jnp.array(.87962566103423978, dtype)
-      return random.truncated_normal(key, -2, 2, shape, dtype) * stddev
+      if jnp.issubdtype(dtype, jnp.floating):
+        # constant is stddev of standard normal truncated to (-2, 2)
+        stddev = jnp.sqrt(variance) / jnp.array(.87962566103423978, dtype)
+        return random.truncated_normal(key, -2, 2, shape, dtype) * stddev
+      else:
+        stddev = jnp.sqrt(variance) / jnp.array(.95311164380491208, dtype)
```
Comment (jekbradbury) [inline jax/_src/nn/initializers.py:129]: Maybe add a comment with the provenance of this constant, as above?

**Review by jekbradbury — APPROVED** [review]

File: jax/_src/nn/initializers.py (hunk 1 above)
Comment (wdphy16) [reply jax/_src/nn/initializers.py:51]: Ok, let me use a one-liner

File: jax/_src/nn/initializers.py (hunk 2 above)
Comment (wdphy16) [reply jax/_src/nn/initializers.py:59]: There was some discussions #4805, and we decided not to add them to `jax.random` for now.

Comment (PhilipVinc) [reply jax/_src/nn/initializers.py:51]: I agree that consensus lacks for `complex_uniform`, though I think that `truncated_normal` could be moved to `jax.random` as there is consensus on what that means for complex numbers.

**Comment by wdphy16** [comment]
Any progress on this?

**Pull request merged**